src/components/SearchInput/SearchInput.tsx

```diff
@@ -0,0 +1,121 @@
+import {
+  AriaAttributes,
+  ChangeEvent,
+  FocusEvent,
+  MouseEvent,
+  ReactNode,
+  useEffect,
+  useRef,
+  useState,
+} from "react"
+import styles from "./SearchInput.module.css"
+
+interface SearchInputProps extends Pick<AriaAttributes, "aria-label"> {
+  name?: string
+  placeholder?: string
+  searchIcon?: ReactNode
+  clearIcon?: ReactNode
+  debounceDelay?: number
+  value?: string
+  onSearch?: (searchTerm: string) => void
+  showSearchButton?: boolean
+  onBlur?: (e: FocusEvent<HTMLInputElement>) => void
+  onFocus?: (e: FocusEvent<HTMLInputElement>) => void
+}
+
+export function SearchInput({
+  name,
+  placeholder,
+  searchIcon,
+  clearIcon,
+  onSearch,
+  debounceDelay = 300,
+  showSearchButton,
+  value = "",
+  onBlur,
+  onFocus,
+  ...aria
+}: SearchInputProps) {
+  const [searchTerm, setSearchTerm] = useState("")
+  const [cancelButtonHidden, setCancelButtonHidden] = useState(true)
+  const timer = useRef<NodeJS.Timeout>()
+
+  useEffect(() => {
+    setSearchTerm(value)
+  }, [value])
+
+  useEffect(() => {
+    if (searchTerm && !showSearchButton) {
+      timer.current = setTimeout(() => onSearch?.(searchTerm), debounceDelay)
+    }
+    return () => {
+      clearTimeout(timer.current)
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [searchTerm, debounceDelay])
+
+  const handleReset = (e: MouseEvent<HTMLButtonElement>) => {
+    // we don't want the field to blur when a user clears the input
+    e.preventDefault()
+    setSearchTerm("")
+    onSearch?.("")
+    setCancelButtonHidden(true)
+  }
+
+  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
+    const currentValue = e.currentTarget.value
+    setSearchTerm(currentValue)
+    setCancelButtonHidden(!currentValue)
+  }
+
+  const handleBlur = (e: FocusEvent<HTMLInputElement>) => {
+    onBlur?.(e)
+    setCancelButtonHidden(true)
+  }
+
+  const handleFocus = (e: FocusEvent<HTMLInputElement>) => {
+    onFocus?.(e)
+    setCancelButtonHidden(!searchTerm.length)
+  }
+
+  return (
+    <div className={styles.wrapper}>
+      <div className={styles.container} data-testid="container">
+        <span data-testid="search_icon" className={styles.searchIcon}>
+          {searchIcon || <span>🔍</span>}
+        </span>
+        <input
```

Comment:
If we don't provide a label for the search field, screenreader users may not know what it's used for. What do you think about adding `ariaLabel` as a required prop as a way to ensure descriptive text?

Reply:
I was considering that and exposing an `id` prop in order to bind a label to the input.